File: system/src/main/java/com/leo/system/AudioFocusHelp.java
/**
 * Copyright (C) 2016 Baidu, Inc. All Rights Reserved.
 */
package com.leo.system;

import android.content.Context;
import android.media.AudioManager;
import android.media.AudioManager.OnAudioFocusChangeListener;

/**
 * 音频焦点辅助类
 */
public class AudioFocusHelp implements OnAudioFocusChangeListener {
    public static final String TAG = AudioFocusHelp.class.getSimpleName();
    public static int AUDIO_STREAM_TYPE = AudioManager.STREAM_MUSIC;
    private static AudioFocusHelp mInstance;
    private AudioManager mAudioManager;
    private OnAudioFocusChangeListener mRegisterAudioFocusListener;
    private int mAudioFocus;

    private AudioFocusHelp() {
        if (mAudioManager == null) {
            mAudioManager = (AudioManager) ContextHelp.getContext()
                    .getSystemService(Context.AUDIO_SERVICE);
        }
    }

    public static AudioFocusHelp getInstance() {
        if (mInstance == null) {
            synchronized (AudioFocusHelp.class) {
                if (mInstance == null) {
                    mInstance = new AudioFocusHelp();
                }
            }
        }
        return mInstance;
    }

    public int requestAudioFocus() {
        if (mAudioManager == null) {
            return AudioManager.AUDIOFOCUS_REQUEST_FAILED;
        }
        if (AudioManager.AUDIOFOCUS_REQUEST_GRANTED == mAudioManager.requestAudioFocus(this,
                AUDIO_STREAM_TYPE, AudioManager.AUDIOFOCUS_GAIN_TRANSIENT)) {
            mAudioFocus = AudioManager.AUDIOFOCUS_GAIN;
            return AudioManager.AUDIOFOCUS_REQUEST_GRANTED;
        } else {
            return AudioManager.AUDIOFOCUS_REQUEST_FAILED;
        }
    }

    public int abandonAudioFocus() {
        if (AudioManager.AUDIOFOCUS_REQUEST_GRANTED == mAudioManager.abandonAudioFocus(this)) {
            mAudioFocus = AudioManager.AUDIOFOCUS_LOSS;
            return AudioManager.AUDIOFOCUS_REQUEST_GRANTED;
        } else {
            return AudioManager.AUDIOFOCUS_REQUEST_FAILED;
        }
    }

    public void setAudioStreamType(int type) {
        this.AUDIO_STREAM_TYPE = type;
    }

    public int getAudioFocus() {
        return mAudioFocus;
    }

    public void registerAudioFocusListener(OnAudioFocusChangeListener mTtsAudioFocusListener) {
        mRegisterAudioFocusListener = mTtsAudioFocusListener;
    }

    @Override
    public void onAudioFocusChange(int focusChange) {
        if (mRegisterAudioFocusListener != null) {
            mRegisterAudioFocusListener.onAudioFocusChange(focusChange);
        }
        mAudioFocus = focusChange;
        switch (focusChange) {
            case AudioManager.AUDIOFOCUS_GAIN:
            case AudioManager.AUDIOFOCUS_GAIN_TRANSIENT:
                break;
            case AudioManager.AUDIOFOCUS_LOSS:
            case AudioManager.AUDIOFOCUS_LOSS_TRANSIENT:
                break;
            default:
                break;
        }
    }
}

File: system/src/main/java/com/leo/system/AudioFocusHelp.kt
/**
 * Copyright (C) 2016 Baidu, Inc. All Rights Reserved.
 */
package com.leo.system

import android.content.Context
import android.media.AudioManager
import android.media.AudioManager.OnAudioFocusChangeListener

/**
 * 音频焦点辅助类
 */
class AudioFocusHelp private constructor() : OnAudioFocusChangeListener {
    private var mAudioManager: AudioManager? = null
    private var mRegisterAudioFocusListener: OnAudioFocusChangeListener? = null
    private var audioFocus: Int = 0

    init {
        if (mAudioManager == null) {
            mAudioManager = ContextHelp.context
                    .getSystemService(Context.AUDIO_SERVICE) as AudioManager
        }
    }

    fun requestAudioFocus(): Int {
        mAudioManager ?: return AudioManager.AUDIOFOCUS_REQUEST_FAILED
        if (AudioManager.AUDIOFOCUS_REQUEST_GRANTED == mAudioManager!!.requestAudioFocus(this,
                        AUDIO_STREAM_TYPE, AudioManager.AUDIOFOCUS_GAIN_TRANSIENT)) {
            audioFocus = AudioManager.AUDIOFOCUS_GAIN
            return AudioManager.AUDIOFOCUS_REQUEST_GRANTED
        } else {
            return AudioManager.AUDIOFOCUS_REQUEST_FAILED
        }
    }

    fun abandonAudioFocus(): Int {
        if (AudioManager.AUDIOFOCUS_REQUEST_GRANTED == mAudioManager!!.abandonAudioFocus(this)) {
            audioFocus = AudioManager.AUDIOFOCUS_LOSS
            return AudioManager.AUDIOFOCUS_REQUEST_GRANTED
        } else {
            return AudioManager.AUDIOFOCUS_REQUEST_FAILED
        }
    }

    fun setAudioStreamType(type: Int) {
        AUDIO_STREAM_TYPE = type
    }

    fun registerAudioFocusListener(mTtsAudioFocusListener: OnAudioFocusChangeListener) {
        mRegisterAudioFocusListener = mTtsAudioFocusListener
    }

    override fun onAudioFocusChange(focusChange: Int) {
        if (mRegisterAudioFocusListener != null) {
            mRegisterAudioFocusListener!!.onAudioFocusChange(focusChange)
        }
        audioFocus = focusChange
        when (focusChange) {
            AudioManager.AUDIOFOCUS_GAIN -> LogUtil.i(TAG, "focusChange：AUDIOFOCUS_GAIN")
            AudioManager.AUDIOFOCUS_GAIN_TRANSIENT -> LogUtil.i(TAG, "focusChange：AUDIOFOCUS_GAIN_TRANSIENT")
            AudioManager.AUDIOFOCUS_LOSS -> LogUtil.i(TAG, "focusChange：AUDIOFOCUS_LOSS")
            AudioManager.AUDIOFOCUS_LOSS_TRANSIENT -> LogUtil.i(TAG, "focusChange：AUDIOFOCUS_LOSS_TRANSIENT")
            else -> {
            }
        }
    }

    companion object {
        private val TAG: String = AudioFocusHelp::class.java.simpleName
        var AUDIO_STREAM_TYPE = AudioManager.STREAM_MUSIC
        private var mInstance: AudioFocusHelp? = null

        fun getInstance(): AudioFocusHelp {
            return mInstance ?: synchronized(this) {
                mInstance ?: AudioFocusHelp().also { mInstance = it }
            }
        }
    }
}
